### PR TITLE
Garbage collection

### PR DESCRIPTION
**Pull request by acdlite**
Adds lifecycle methods that are invoked during the complete phase. These can be used to prepare external resources, like subscriptions, before work is committed.

Corresponding lifecycle methods are invoked if work is dropped without committing. Otherwise, the normal commit phase lifecycles (`componentDidMount`, `componentDidUpdate`) are called.

In theory, the same underlying mechanism could be used to create and destroy persistent Yoga views in React Native without leaking memory.

The complete phase lifecycles are:
- `unstable_prepareMount`
- `unstable_prepareUpdate`

The garbage collection lifecycles are:
- `unstable_abortMount`
- `unstable_abortUpdate`

To avoid memory leaks, every call of a "prepare" lifecycle must have either a corresponding commit phase lifecycle or a corresponding "abort" lifecycle.

TODO:
- [ ] Dropped trees don't need to be garbage collected immediately. Defer by scheduling a callback. Requires refactoring to give `createWorkInProgress` access to the scheduler.

Based on top of my other open async PRs. Relevant work starts with the commit titled "Garbage collection."

### CODE REVIEW

File: src/renderers/shared/fiber/ReactFiberCompleteWork.js [hunk 1]
```diff
@@ -198,7 +206,33 @@ module.exports = function<T, P, I, TI, PI, C, CX, PL>(
       case FunctionalComponent:
         return null;
       case ClassComponent: {
-        // We are leaving this subtree, so pop context if any.
+        if (
+          workInProgress.effectTag & PerformedWork &&
+          !(workInProgress.effectTag & PreparedWork)
```
Comment (clemmy) [inline src/renderers/shared/fiber/ReactFiberCompleteWork.js]: In what scenarios can both `workInProgress.effectTag & PerformedWork` and `workInProgress.effectTag & PreparedWork` be true?

Comment (acdlite) [reply src/renderers/shared/fiber/ReactFiberCompleteWork.js]: Names are a bit confusing. `PerformedWork` means this fiber was worked on. `PreparedWork` means this fiber has stuff that needs to be garbage collected if it's dropped without committing.

Comment (acdlite) [reply src/renderers/shared/fiber/ReactFiberCompleteWork.js]: They're separate because not all fibers that are worked on need to be garbage collected.

**Comment by gaearon** [comment]
Stality stale.

**Pull request closed without merging**